### ISSUE TITLE
Performance of sk_single experiment

### ISSUE THREAD

**Issue by Eaphan** [valeoai/UNIT#5]
I use the default config to train a sk_single model, but the performance is not as good as the values paper claimed.

command:
python main_instance_segmentation.py general.experiment_name=sk_single data/datasets=semantic_kitti data.batch_size=3 general.num_frames=1 data.test_batch_size=3 trainer.max_epochs=50

Num of gpus: 8

Evaluation results:
Scores with no filtering
S_assoc (LAQ): 0.40612851961510066
S_assoc_temporal (LAQ): 0.06872678306565637
IoU* (LAQ): 0.1499578289944978
Scores with 50 points filtering
S_assoc (LAQ): 0.5048753466097594
S_assoc_temporal (LAQ): 0.08063252794017975
IoU* (LAQ): 0.1716526305656636

I also evaluated your released model; the performance is close to that in the paper.

**Comment by Eaphan** [comment]
And when I use your released sk_single.ckpt as the initial weights of the second training stage, the temporal model achieves similar performance to the paper.

S_assoc (LAQ): 0.693449147631201
S_assoc_temporal (LAQ): 0.42253554346821653
IoU* (LAQ): 0.5152432381057271
Scores with 50 points filtering
S_assoc (LAQ): 0.7836535239142897
S_assoc_temporal (LAQ): 0.4918915182322448
IoU* (LAQ): 0.5851417759954171

**Comment by CSautier** [comment]
I don't really know what could have gone wrong. In my recollection, the "single" training was actually more stable. Have you looked as your pre-processing, to make sure that nothing is wrong here?

**Comment by Eaphan** [comment]
"For SemanticKITTI, you can directly [download](https://github.com/valeoai/UNIT/releases/download/v1.0/segments_gridsample_sk.tar.gz) the pre-processing as re-running Patchwork on your side can be complicated. "


I directly use the segments you provided.

**Comment by Eaphan** [comment]
python main_instance_segmentation.py general.experiment_name=sk_single \
    data/datasets=semantic_kitti data.batch_size=3 general.num_frames=1 \
    data.test_batch_size=3 trainer.max_epochs=50

python inference_instance_segmentation.py general.experiment_name=sk_single general.checkpoint=saved/sk_single/last-epoch.ckpt data.batch_size=1 general.forward_queries=true data/datasets=semantic_kitti data.predict_mode=validation

These are the commands I used for training and inference.

And I train the model with 4/8 NVIDIA 3090 gpus.

**Comment by Eaphan** [comment]
I'm trying to add the option "general.precision=32" to the options to deactivate AMP training.

**Comment by CSautier** [comment]
The batch size was kept very low to train on a single GPU if I remember correctly. I don't remember if I had experimented successfully with multiple GPUs

**Comment by Eaphan** [comment]
It would be great if you could try it on multiple gpus as well.

**Comment by Eaphan** [comment]
> It would be great if you could try it on multiple gpus as well.

It seems I obtain normal performance when I use only one gpu for training. Next, I will try to change the code to support multi-gpu training.

**Comment by CSautier** [comment]
When you train on 4 GPUs, given that the batch size is of 3 and not divisible by 4, I'm not sure what actually happens. Most likely, 3 is the batch size _per gpu_, meaning that the effective batch size is 12, and then the total number of training iterations is 4 times lesser than in the recommended training. I would recommend then to increase the learning rate, and maybe also the number of epochs. I don't have the resources at the moment to test such a training.

**Comment by Eaphan** [comment]
Currently,  the codebase of Mask3D does not support multi-gpu training.

https://github.com/JonasSchult/Mask3D/issues/26

**Comment by CSautier** [comment]
This specific bug you linked is a common bug with MinkowskiEngine, which does not seems to be the exact problem you are facing. I'm thinking that the performance drop you experience might more be caused by a problem in either learning rate, number of iterations, or maybe if the loss is actually only computed in one of the 4 GPUs.

Neither Mask3D nor UNIT were conceived with multiGPU in mind, especially since it goes against the idea of an online tracker, but I think it could be made to work.
